### PR TITLE
Adding Azure Cosmos Mongo vCore Vector DB Cache

### PR DESCRIPTION
**Pull request by aayush3011**
Description:

This pull request introduces several enhancements for Azure Cosmos Vector DB, primarily focused on improving caching and search capabilities using Azure Cosmos MongoDB vCore Vector DB. Here's a summary of the changes:

 - **AzureCosmosDBSemanticCache**: Added a new cache implementation called AzureCosmosDBSemanticCache, which utilizes Azure Cosmos MongoDB vCore Vector DB for efficient caching of semantic data. Added comprehensive test cases for AzureCosmosDBSemanticCache to ensure its correctness and robustness. These tests cover various scenarios and edge cases to validate the cache's behavior.
 - **HNSW Vector Search**: Added HNSW vector search functionality in the CosmosDB Vector Search module. This enhancement enables more efficient and accurate vector searches by utilizing the HNSW (Hierarchical Navigable Small World) algorithm. Added corresponding test cases to validate the HNSW vector search functionality in both AzureCosmosDBSemanticCache and AzureCosmosDBVectorSearch. These tests ensure the correctness and performance of the HNSW search algorithm.
 - **LLM Caching Notebook** - The notebook now includes a comprehensive example showcasing the usage of the AzureCosmosDBSemanticCache. This example highlights how the cache can be employed to efficiently store and retrieve semantic data. Additionally, the example provides default values for all parameters used within the AzureCosmosDBSemanticCache, ensuring clarity and ease of understanding for users who are new to the cache implementation.
 
 @hwchase17,@baskaryan, @eyurtsev,

### CODE REVIEW

File: libs/community/langchain_community/vectorstores/azure_cosmos_db.py [hunk 1]
```diff
@@ -409,17 +538,93 @@ def max_marginal_relevance_search_by_vector(
...
-            embeddings, k=k, fetch_k=fetch_k, lambda_mult=lambda_mult
+            embeddings, k=k, kind=kind, fetch_k=fetch_k, lambda_mult=lambda_mult, score_threshold=score_threshold
         )
         return docs
+
+    def get_collection(self):
+        return self._collection
+
+
+class AzureCosmosDBVectorSearchRetriever(BaseRetriever):
```
Comment (baskaryan) [inline libs/community/langchain_community/vectorstores/azure_cosmos_db.py:562]: how's this class different from VectorStoreRetriever?

Comment (aayush3011) [reply libs/community/langchain_community/vectorstores/azure_cosmos_db.py:562]: I have removed the AzureCosmosDBVectorSearchRetriever class, as it is identical implementation to the VectorStoreRetriever class.

**Comment by aayush3011** [comment]
Can you please review this PR, @hwchase17,@baskaryan, @eyurtsev.

**Comment by aayush3011** [comment]
@baskaryan , can you please take another look and we can merge the PR by EOW.

File: libs/community/langchain_community/vectorstores/azure_cosmos_db.py [hunk 2]
```diff
@@ -163,9 +177,12 @@ def delete_index(self) -> None:
 
     def create_index(
         self,
+        kind: str = "vector-ivf",
         num_lists: int = 100,
-        dimensions: int = 1536,
         similarity: CosmosDBSimilarityType = CosmosDBSimilarityType.COS,
+        dimensions: int = 1536,
+        m: int = 16,
+        ef_construction: int = 64,
```
Comment (baskaryan) [inline libs/community/langchain_community/vectorstores/azure_cosmos_db.py:185]: switching the order of these is a breaking change. can we keep (self, num_lists, dimensions, similarity) in that order and add all new args at the end?

Comment (aayush3011) [reply libs/community/langchain_community/vectorstores/azure_cosmos_db.py]: @baskaryan, made the changes as suggested, and fixed the linting issues as well.

**Comment by aayush3011** [comment]
@baskaryan can you please trigger the pipelines again, there were lint failures.

**Comment by aayush3011** [comment]
@baskaryan can you please trigger the pipelines again, there were lint failures.

**Comment by aayush3011** [comment]
@baskaryan can you please trigger the pipelines again, there were mypy failures.

I have corrected which seemed like actual issues.

**Comment by aayush3011** [comment]
@baskaryan can you please trigger the pipelines again. This should be the last time :)

We can merge the PR after this iteration.

**Comment by aayush3011** [comment]
@baskaryan can you please trigger the pipelines again.

**Comment by aayush3011** [comment]
Hey @baskaryan, can we please try and close out this PR today. It has been open for a few weeks.

**Review by baskaryan — APPROVED** [review]

**Comment by aayush3011** [comment]
@baskaryan can you please trigger the pipelines again, and can you merge after this iteration runs.

**Comment by aayush3011** [comment]
@baskaryan can you please trigger the pipelines again, and can we merge after this iteration runs.

**Comment by aayush3011** [comment]
@baskaryan can you please trigger the pipelines again. I want to get this in today. Can you please be online for some time, just in case if the pipelines fail again.

**Comment by aayush3011** [comment]
@baskaryan, can we go ahead and merge the PR now.
